### PR TITLE
Feat/bucket internal impl

### LINKED ISSUE #236: Implement CRUD Methods for Buckets
### Code Reference

```
rpc CreateBucket(MsgCreateBucket) returns (MsgCreateBucketResponse);
```

```
rpc ReadBucket(MsgReadBucket) returns (MsgReadBucketResponse);
```

```
rpc UpdateBucket(MsgUpdateBucket) returns (MsgUpdateBucketResponse);
```

```
rpc DeleteBucket(MsgDeleteBucket) returns (MsgDeleteBucketResponse);
```

```
rpc CreateWhichIs(MsgCreateWhichIs) returns (MsgCreateWhichIsResponse);
```

```
rpc UpdateWhichIs(MsgUpdateWhichIs) returns (MsgUpdateWhichIsResponse);
```

```
rpc DeleteWhichIs(MsgDeleteWhichIs) returns (MsgDeleteWhichIsResponse);
```

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202528620406598/1202529474434550) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by joshLong145**
* Addition of internal bucket module for motor
## Changes
* Addition of `bucket` in `internal`
## API Updates
* Addition of `Bucket` interface which will outline the `motor` api.

```
type Bucket interface {
	/*
		Retrieves a piece of content by the given uri
	*/
	GetContentById(id string) (*BucketContent, error)

	GetBucketItems() []*bt.BucketItem

	/*
		Checks if a given uri for existence in the given bucket
	*/
	ContentExists(id string) bool

	/*
		Checks if a given uri is for a bucket
	*/
	IsBucket(id string) bool

	/*
		Checks if a given uri is for an object
	*/
	IsContent(id string) bool

	/*
		Resolves all buckets defined within the current WhereIs by `did`
	*/
	ResolveBuckets(address string) error

	/*
		Resolves all content within the bucket by `cid`
	*/
	ResolveContent() error
}
```

### New Features _(required)_
* Ability to Get content from a bucket
  * Objects
  * Other Buckets 
* [ ] Unit tests
* [ ] Documentation

<sub> Mention [stepsize] in a comment if you'd like to report some technical debt. See examples [here](https://app.stepsize.com/api/demo-pr-redirect).</sub>

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202528620406598/1202810799820249) by [Unito](https://www.unito.io)

### CODE REVIEW

File: internal/bucket/bucket.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	errContentNotFound = func(id string) error {
+		if id != "" {
+			return fmt.Errorf("could not find content with id: %s", id)
+		}
+
+		return errors.New("could not find content")
+	}
+)
+
+type Bucket interface {
```
Comment (ntindle) [inline internal/bucket/bucket.go]: I feel like we should have more here but not sure why

Comment (ntindle) [reply internal/bucket/bucket.go]: List content without resolving?

**Review by ntindle — COMMENTED** [review]
How might I update a bucket with this

File: internal/bucket/bucket.go (hunk 1 above)
Comment (joshLong145) [reply internal/bucket/bucket.go]: thats a good idea, ill add `GetContent() []*BucketContent`

**Review by ntindle — APPROVED** [review]
LGTM

File: internal/bucket/resolve.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		return errors.New("top level bucket not provided")
+	}
+	var buckets []string = make([]string, 0)
+	for _, b := range b.whereIs.Content {
+		if b.Type == types.ResourceIdentifier_DID {
+			buckets = append(buckets, b.Uri)
+		}
+	}
+
+	for len(buckets) > 0 {
```
Comment (mcjcloud) [inline internal/bucket/resolve.go]: Why remove elements from the list every iteration? I think the two loops can be combined where you just `continue` past an element if it's not a DID.

This applies to `ResolveContent` too.

File: internal/bucket/verify.go [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package bucket
+
+func (b *bucketImpl) IsBucket(id string) bool {
```
Comment (mcjcloud) [inline internal/bucket/verify.go:5]: unimplemented?

**Review by mcjcloud — CHANGES_REQUESTED** [review]

File: internal/bucket/resolve.go (hunk 2 above)
Comment (joshLong145) [reply internal/bucket/resolve.go]: I modified the implementation to only resolve the top level, so yes there is no need for the other loop

**Review by mcjcloud — APPROVED** [review]

**Review by prdlk — APPROVED** [review]

**Pull request merged**